File: solutions/15. 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        vector<vector<int>> res;
        int ptr1 = 0;
        int low = 0;
        int high = 0;
        sort(nums.begin(), nums.end());
        for(int i=0;i<nums.size();i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            ptr1 = i;
            low = i+1;
            high = nums.size()-1;
            while(low<high){
                
                int sum = nums[ptr1]+nums[low]+nums[high];
                if(sum>0){
                    high--;
                } else if(sum<0){
                    low++;
                } else {
                    res.push_back({nums[ptr1], nums[low], nums[high]});
                    low++;
                    high--;
                    while(low<high && nums[high]==nums[high+1]) high--;
                    while(low<high && nums[low]==nums[low-1]) low++;
                }
            }
        }
        return res;
    }
};

File: solutions/Array/15. 3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        if(nums.length<3){
            return new ArrayList<>();
        }
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for(int i = 0;i<nums.length-2;i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int fix = nums[i], start = i+1, end = nums.length-1, target = 0;
            while(start<end){
                int sum = fix + nums[start] + nums[end];
                if(sum>target){
                    end--;
                } else if (sum<target){
                    start++;
                } else {
                    res.add(Arrays.asList(fix,nums[start],nums[end]));
                    end--;
                    start++;
                    while(start<end && nums[start]==nums[start-1]) start++;
                    while(start<end && nums[end]==nums[end+1]) end--;
                }
            }
        }
        return res;
    }
}